From f7b9e80ef7c59bd834c892b1e67d7fc8b0255060 Mon Sep 17 00:00:00 2001
From: Simon Shields <simon@lineageos.org>
Date: Tue, 29 Aug 2017 23:03:29 +1000
Subject: [PATCH 13/30] edify: bring back SetPermissionsRecursive

* we still need this, even with block-based OTAs.

This partially reverts commit f388104eaacd05cfa075d6478369e1d0df5ddbf3

Change-Id: I4943e2e89ee5c810a63746c570dc5e31e95b8c53
---
 core/Makefile                         |  1 +
 tools/releasetools/edify_generator.py | 17 +++++++++++++++++
 2 files changed, 18 insertions(+)

diff --git a/core/Makefile b/core/Makefile
index 8796c2e..f6ea96b 100644
--- a/core/Makefile
+++ b/core/Makefile
@@ -2355,6 +2355,7 @@ ifdef PRODUCT_EXTRA_RECOVERY_KEYS
 endif
 	$(hide) echo 'mkbootimg_args=$(BOARD_MKBOOTIMG_ARGS)' >> $(zip_root)/META/misc_info.txt
 	$(hide) echo 'mkbootimg_version_args=$(INTERNAL_MKBOOTIMG_VERSION_ARGS)' >> $(zip_root)/META/misc_info.txt
+	$(hide) echo "use_set_metadata=1" >> $(zip_root)/META/misc_info.txt
 	$(hide) echo "multistage_support=1" >> $(zip_root)/META/misc_info.txt
 	$(hide) echo "blockimgdiff_versions=3,4" >> $(zip_root)/META/misc_info.txt
 ifneq ($(OEM_THUMBPRINT_PROPERTIES),)
diff --git a/tools/releasetools/edify_generator.py b/tools/releasetools/edify_generator.py
index e9ed798..a768bc5 100644
--- a/tools/releasetools/edify_generator.py
+++ b/tools/releasetools/edify_generator.py
@@ -294,6 +294,23 @@ class EdifyGenerator(object):
     cmd = "".join(cmd)
     self.script.append(self.WordWrap(cmd))
 
+  def SetPermissionsRecursive(self, fn, uid, gid, dmode, fmode, selabel,
+                              capabilities):
+    """Recursively set path ownership and permissions."""
+    if not self.info.get("use_set_metadata", False):
+      self.script.append('set_perm_recursive(%d, %d, 0%o, 0%o, "%s");'
+                         % (uid, gid, dmode, fmode, fn))
+    else:
+      if capabilities is None:
+        capabilities = "0x0"
+      cmd = 'set_metadata_recursive("%s", "uid", %d, "gid", %d, ' \
+          '"dmode", 0%o, "fmode", 0%o, "capabilities", %s' \
+          % (fn, uid, gid, dmode, fmode, capabilities)
+      if selabel is not None:
+        cmd += ', "selabel", "%s"' % selabel
+      cmd += ');'
+      self.script.append(cmd)
+
   def WriteRawImage(self, mount_point, fn, mapfn=None):
     """Write the given package file into the partition for the given
     mount point."""
-- 
2.7.4

